### PR TITLE
Avoid panics when using HTMLAnchorElement attribute setters

### LINKED ISSUE #10877: HTMLAnchorElement search setter panics due to borrow error
```html
<a>anchor</a>
<script>
var a = document.querySelector('a');
a.href = "http://google.com?ho";
a.search = "?hi";
</script>
```

```
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #7: tid = 0x27b1b8, 0x000000010325c1f0 servo`rust_panic, stop reason = breakpoint 1.1
  * frame #0: 0x000000010325c1f0 servo`rust_panic
    frame #1: 0x0000000103251d13 servo`std::sys_common::unwind::begin_unwind_inner::h30e12d15ce2b2e25 + 467
    frame #2: 0x00000001032535bf servo`std::sys_common::unwind::begin_unwind_fmt::hb2de8a9968d38523 + 207
    frame #3: 0x00000001032636c8 servo`rust_begin_unwind + 56
    frame #4: 0x0000000103291aa1 servo`core::panicking::panic_fmt::h257ceb0aa351d801 + 129
    frame #5: 0x0000000103299566 servo`core::option::expect_failed::h2d57a5644f345e0b + 102
    frame #6: 0x000000010162bac4 servo`core::option::{{impl}}::expect<core::cell::Ref<core::option::Option<url::Url>>>(self=Option<core::cell::Ref<core::option::Option<url::Url>>> at 0x000000011abf08f8, msg=&str at 0x000000011abf08e8) + 180 at option.rs:293
    frame #7: 0x000000010162ba02 servo`script::dom::bindings::cell::{{impl}}::borrow<core::option::Option<url::Url>>(self=0x000000011ae7c310) + 66 at cell.rs:136
    frame #8: 0x000000010162bd94 servo`script::dom::htmlanchorelement::{{impl}}::update_href(self=0x000000011ae7c1c0) + 100 at htmlanchorelement.rs:86
    frame #9: 0x00000001010db8cc servo`script::dom::htmlanchorelement::{{impl}}::SetSearch(self=0x000000011ae7c1c0, value=USVString at 0x000000011abf0ac8) + 316 at htmlanchorelement.rs:411
    frame #10: 0x00000001010db732 servo`script::dom::bindings::codegen::Bindings::HTMLAnchorElementBinding::set_search(cx=0x0000000108e89900, obj=Handle<*mut js::jsapi_macos_64::JSObject> at 0x000000011abf0bb8, this=0x000000011ae7c1c0, args=JSJitSetterCallArgs at 0x000000011abf0ba8) + 338 at HTMLAnchorElementBinding.rs:886
    frame #11: 0x0000000102b6e7fb servo`CallJitSetterOp + 187
    frame #12: 0x0000000100f11cd5 servo`script::dom::bindings::utils::call_setter(info=0x0000000104ad3a40, cx=0x0000000108e89900, handle=Handle<*mut js::jsapi_macos_64::JSObject> at 0x000000011abf0cd0, this=0x000000011ae7c1c0, argc=1, vp=0x000000011abf1858) + 85 at utils.rs:524
```

### PR DESCRIPTION
**Pull request by broesamle**
`expected: OK` still explicitly mentioned in .ini file. Shall I remove it?

Fixes #10877.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10903)

### CODE REVIEW

File: tests/wpt/mozilla/meta/mozilla/htmlanchorelment-search-attribute.html.ini [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[htmlanchorelment-search-attribute.html]
+  type: testharness
+  expected: OK
```
Comment (KiChjang) [inline tests/wpt/mozilla/meta/mozilla/htmlanchorelment-search-attribute.html.ini]: Yes, remove this. By default, tests are expected to PASS (or OK), so if there's no .ini file, the test framework will expect the test file to pass.

**Comment by broesamle** [comment]
fixes #10877

**Comment by KiChjang** [comment]
Speaking of which, why isn't this a WPT test?

File: tests/wpt/mozilla/tests/mozilla/htmlanchorelment-search-attribute.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<!doctype html>
+<meta charset="utf-8">
+<html>
+<head>
+<title>HTMLAnchorElement search setter crashes</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<a>anchor</a>
```
Comment (Ms2ger) [inline tests/wpt/mozilla/tests/mozilla/htmlanchorelment-search-attribute.html]: Please put this in the directory of upstream tests, add `assert_*`s that the setter works correctly, and also test all the other setters you changed.

Comment (broesamle) [inline tests/wpt/mozilla/tests/mozilla/htmlanchorelment-search-attribute.html]: @jdm points to an overlap regarding the tests with #10876 and suggests:
> Test: We should create a test that exercises each of the getters and setters individually in tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/.

Shall I just move it there?

Comment (broesamle) [inline tests/wpt/mozilla/tests/mozilla/htmlanchorelment-search-attribute.html]: to rephrase my question (making me hesitate to just move the test):
Is there a specific location where to put it, before it formally enters the WPT repo?

what you called:
> the directory of upstream tests

Comment (Ms2ger) [inline tests/wpt/mozilla/tests/mozilla/htmlanchorelment-search-attribute.html]: `tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/` is perfect.

**Comment by broesamle** [comment]
well, I tried to merge the the upstream(servo/servo) master into a local branch of my fork. 

The conflict in MANIFEST.json looks a bit awkward to resolve manually :-/
"Rebuilding" MANIFEST.json could be easier: 
`./mach test-wpt --manifest-update` 
Would that work, does that make sense?

**Comment by jdm** [comment]
It would, but I recommend `./mach test-wpt --manifest-update SKIP_TESTS` to avoid running any tests.

**Comment by broesamle** [comment]
done the rebase...at least the conflict is gone :-)

**Comment by broesamle** [comment]
Manually merged MANIFEST.json
all tests that I touched can be run (so the MANIFEST should be ok), despite some did not pass.

**Comment by broesamle** [comment]
`The command "bash etc/ci/manifest_changed.sh" exited with 1.`
:-( 

so, what is the best practice of resolving merge conflicts from MANIFEST.json?

**Comment by jdm** [comment]
Huh, it should be enough to undo any changes to MANIFEST.json and run `./mach test-wpt --manifest-update SKIP_TESTS`.

**Comment by broesamle** [comment]
let's see ... this time I used the automatic update again.

**Comment by broesamle** [comment]
you were right, @jdm , I should not have merged MANIFEST.json by hand.

**Comment by jdm** [comment]
Looking good! I think we can make the new test easier to read and modify, however.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 2 files at r2, 1 of 3 files at r4, 2 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 11 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10903#-KGrRsP59vq-bUv3uxb-:-KGrRsP59vq-bUv3uxb0:-1558327943) ([raw file](https://github.com/servo/servo/blob/744d665b286985b5c9d3a9ec253ebb9506748944/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L11)):*
There's a lot of duplication in each of these tests. I think we can do something like this:
```js
function test_setter(property, old, new) {
  /* initialize a to known url */
  assert_equals(a[property], old);
  a[property] = new;
  assert_equals(a[property], new);
}

var tests = [["hash", "#somehash", "#someother"], ...];
for (var i = 0; i < tests.length; i++) {
  test(function() {
    test_setter(tests[i][0], tests[i][1], tests[i][2]);
  }, "Getter and setter for attribute of anchor element: " + tests[i][0])
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10903#-:-KGrT5v0MRr1gSUs8wXb:-632064963)*

**Comment by broesamle** [comment]
agreed. I rewrote the test according to your suggestion -- with the slight 'complexication' that I have an additional check for the old and new url.

Which already produced an unexpected (to me) fail in the search setter

`assert_equals: expected "http://google.com?hi" but got "http://google.com/?hi"`

Is it valid, if the setter adds the / after the hostname, here? 
(From my quick look in the specs I did not come to a conclusion.)

otherwise all tests pass nicely.

**Comment by broesamle** [comment]
from my POV there is not much to do. please let me know if there is anything more to change and when would be a good time to rebase and then merge it.

**Comment by jdm** [comment]
I apologize for the delay; I'll finish reviewing this today.

**Comment by jdm** [comment]
This looks really good!
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> I apologize for the delay; I'll finish reviewing this today.</details>


---

Reviewed 1 of 4 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 10 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHRdI9VEThtQyCiBKya:-KHRdBvBLlBrKJm0hzSl:-834995750) ([raw file](https://github.com/servo/servo/blob/aa4ff60ea90a454e83900fe6678bae72c40b5a58/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L10)):*
> ```HTML
> <a>anchor</a>
> <script>
> function test_getter(property,url,expectedresult) {
> ```

This function doesn't appear to do anything.

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 17 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHRdLgI1t8jo7McmcJz:-KHRdLgI1t8jo7McmcK-:1147259612) ([raw file](https://github.com/servo/servo/blob/aa4ff60ea90a454e83900fe6678bae72c40b5a58/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L17)):*
> ```HTML
>   a.href = oldurl;
>   var r1 = a[property];
>   assert_equals(r1,oldresult);
> ```

nit: space after `,`

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 20 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHRdOh68QdKE7PSyBIU:-KHRdOh68QdKE7PSyBIV:1147259612) ([raw file](https://github.com/servo/servo/blob/aa4ff60ea90a454e83900fe6678bae72c40b5a58/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L20)):*
> ```HTML
>   a[property] = newval;
>   var r2 = a[property];
>   assert_equals(r2,newresult);
> ```

nit: space after `,`

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 22 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHRdNE1tkjrXb_dQfqB:-KHRdNE1tkjrXb_dQfqC:1147259612) ([raw file](https://github.com/servo/servo/blob/aa4ff60ea90a454e83900fe6678bae72c40b5a58/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L22)):*
> ```HTML
>   assert_equals(r2,newresult);
>   var r3 = a.href;
>   assert_equals(r3,newurl);
> ```

nit: space after `,`

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 57 \[r7\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHReCrPZU52kbPLu1ju:-KHReCrPZU52kbPLu1jv:-93960354) ([raw file](https://github.com/servo/servo/blob/aa4ff60ea90a454e83900fe6678bae72c40b5a58/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L57)):*
> ```HTML
>   ["search", "?ho", "?hi", "?hi", "http://google.com/?ho", "http://google.com/?hi"],
>   ["search", "?ho", "hi", "?hi", "http://google.com/?ho", "http://google.com/?hi"],
> // these don't
> ```

Let's uncomment them and mark them as known failures using an ini file (see [docs](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations)), then file an issue to fix it :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10903#-:-KHReh2Su4ro8rzE1jcr:-1273308284)*

**Comment by broesamle** [comment]
thank you for the quick reply, yesterday!
I'll adapt as suggested as soon as I fixed my inconsistent building system...

**Comment by jdm** [comment]
-S-needs-rebase -S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, broesamle wrote…</i></summary>
> thank you for the quick reply, yesterday!
> 
> I'll adapt as suggested as soon as I fixed my inconsistent building system...</details>


---

Reviewed 1 of 4 files at r5, 1 of 1 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 55 \[r9\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHW5DE62JSd93U1RlVQ:-KHW5DE62JSd93U1RlVR:-1315658782) ([raw file](https://github.com/servo/servo/blob/864d4162ac24281e7ea8832865c82b15865efc47/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L55)):*
> ```HTML
>   ["search", "?ho", "hi", "?hi", "http://google.com/?ho", "http://google.com/?hi"],
>   // FIXME: the following two tests fail
>   // (please check specification what is the correct behaviour if ? directly follows the host without file or /
> ```

The specification says that "http://google.com?ho" is a valid URL, and when serialized it will always have a `/` following the host (https://url.spec.whatwg.org/#concept-url-serializer).

---

*[tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html, line 57 \[r9\]](https://reviewable.io:443/reviews/servo/servo/10903#-KHW5NSIKOmSARgBUa2W:-KHW5NSIKOmSARgBUa2X:-1868492866) ([raw file](https://github.com/servo/servo/blob/864d4162ac24281e7ea8832865c82b15865efc47/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/htmlanchorelement_attribute-getter-setter.html#L57)):*
> ```HTML
>   // (please check specification what is the correct behaviour if ? directly follows the host without file or /
>   ["search", "?ho", "?hi", "?hi", "http://google.com?ho", "http://google.com?hi"],
>   ["search", "?ho", "hi", "?hi", "http://google.com?ho", "http://google.com?hi"],
> ```

The expected output URLs should be `http://google.com/?hi`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10903#-:-KHW5j9eRLHVY1BDgptt:-628972244)*

**Comment by broesamle** [comment]
ok, then they should all pass...

**Comment by broesamle** [comment]
There is a reparse of the url mentioned so you are right.
I removed the failing tests again, because I think now that they are confusing.

Apoligies for the to and fro in this respect ... I was unsure about the specs.

**Comment by jdm** [comment]
Great! Please [squash](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) the commits into one, and I'll merge them!

**Comment by broesamle** [comment]
oh ... just one thing ... should I have updated MANIFEST.json again when squashing?
(I was not too happy with the diff)

**Comment by jdm** [comment]
It looks like it might be using an older version of MANIFEST.json. I recommend running `./mach test-wpt --manifest-update SKIP_TESTS`, amending your commit with the result, and force pushing that to this branch.

**Comment by broesamle** [comment]
That did not change it.

I think it is because my fork is far behind the upstream so that the diff looks so complex. 
One way would be to 
1) remove the manifest from this commit
2) merge it anyway (the test will just be ignored)
3) then rebase my fork with the upstream
4) update and commit a new manifest with the new test

**Comment by broesamle** [comment]
did it. (forget my former concern) ... its late here in berlin.

**Comment by jdm** [comment]
@bors-servo: r+
Thank you very much!

**Comment by broesamle** [comment]
Thank you for your patience! I learned a lot about servo on the way :-)

**Comment by mbrubeck** [comment]
@bors-servo retry

* network error

**Pull request merged**